openedx/core/djangoapps/celery_utils/management/commands/tests/test_reapply_tasks.py

```diff
@@ -0,0 +1,118 @@
+
+    @classmethod
+    def setUpClass(cls):
+        @celery.task(base=task.PersistOnFailureTask, name=cls.fallible_task_name)
+        def fallible_task(error_message=None):
+            """
+            Simple task to let us test retry functionality.
+            """
+            if error_message:
+                raise ValueError(error_message)
+
+        cls.fallible_task = fallible_task
+
+        @celery.task(base=task.PersistOnFailureTask, name=cls.passing_task_name)
+        def passing_task():
+            """
+            This task always passes
+            """
+            return 5
+        cls.passing_task = passing_task
+        super(TestReapplyTaskCommand, cls).setUpClass()
+
+    def setUp(self):
+        self.failed_tasks = [
+            models.FailedTask.objects.create(
+                task_name=self.fallible_task_name,
+                task_id=u'fail_again',
+                args=[],
+                kwargs={"error_message": "Err, yo!"},
+                exc=u'UhOhError().',
+            ),
+            models.FailedTask.objects.create(
+                task_name=self.fallible_task_name,
+                task_id=u'will_succeed',
+                args=[],
+                kwargs={},
+                exc=u'NetworkErrorMaybe?()',
+            ),
+            models.FailedTask.objects.create(
+                task_name=self.passing_task_name,
+                task_id=u'other_task',
+                args=[],
+                kwargs={},
+                exc=u'RaceCondition()',
+            ),
+        ]
+        super(TestReapplyTaskCommand, self).setUp()
+
+    def _assert_resolved(self, task_object):
+        """
+        Raises an assertion error if the task failed to complete successfully
+        and record its resolution in the failedtask record.
+        """
+        self.assertIsInstance(task_object.datetime_resolved, datetime)
+
+    def _assert_unresolved(self, task_object):
+        """
+        Raises an assertion error if the task completed successfully.
+        The resolved_datetime will still be None.
+        """
+        self.assertIsNone(task_object.datetime_resolved)
+
+    def test_call_command(self):
+        call_command(u'reapply_tasks')
+        self._assert_unresolved(models.FailedTask.objects.get(task_id=u'fail_again'))
+        self._assert_resolved(models.FailedTask.objects.get(task_id=u'will_succeed'))
+        self._assert_resolved(models.FailedTask.objects.get(task_id=u'other_task'))
+
+    def test_call_command_with_specified_task(self):
+        call_command(u'reapply_tasks', u'--task-name={}'.format(self.fallible_task_name))
+        self._assert_unresolved(models.FailedTask.objects.get(task_id=u'fail_again'))
+        self._assert_resolved(models.FailedTask.objects.get(task_id=u'will_succeed'))
+        self._assert_unresolved(models.FailedTask.objects.get(task_id=u'other_task'))
+
+    def test_duplicate_tasks(self):
```

Comment:
Can you explain why this is a scenario that must be handled instead of prevented?
I thought the checks you added in `on_success` and `on_failure` will prevent this from happening.  Or is this a side-effect of the existing mysterious issue with having multiple workers running the same task?